common/src/main/scala/explore/model/RootModel.scala

```diff
@@ -1,23 +1,92 @@
-// Copyright (c) 2016-2019 Association of Universities for Research in Astronomy, Inc. (AURA)
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
 // For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
 
 package explore.model
 
+import cats.implicits._
 import cats.effect._
-
 import crystal._
+import clue.js.AjaxGraphQLClient
+import clue.js.WebSocketGraphQLClient
+import explore.graphql.TestQuery
+import io.lemonlabs.uri.Url
 import monocle.macros.Lenses
-
-import scala.concurrent.ExecutionContext.global
+import cats.effect.concurrent.Ref
+import japgolly.scalajs.react.Reusability
 
 @Lenses
 case class RootModel(
-  target: Target
+  target:   Option[Target]             = None,
+  persons:  List[TestQuery.AllPersons] = List.empty,
+  todoList: List[Task]                 = List.empty,
+  polls:    List[Poll]                 = List.empty
 )
+object RootModel {
+  implicit val filmsReuse: Reusability[TestQuery.AllPersons.Films] = Reusability.derive
+  implicit val allPersonsReuse: Reusability[TestQuery.AllPersons]  = Reusability.derive
+  implicit val reuse: Reusability[RootModel]                       = Reusability.derive
+}
+
+case class AppConfig(
+  swapiURL: Url = Url.parse("https://api.graph.cool/simple/v1/swapi"),
+  todoURL: Url = Url.parse(
+    "https://cors-anywhere.herokuapp.com/https://todo-mongo-graphql-server.herokuapp.com/"
+  ),
+  pollURL: Url = Url.parse("wss://realtime-poll.demo.hasura.app/v1/graphql")
+)
+
+class AppState[F[_]: ConcurrentEffect: Timer](
+  config:        AppConfig,
+  val rootModel: Model[F, RootModel]
+) {
+  object Clients {
+    val starWars = AjaxGraphQLClient(config.swapiURL)
+    val todo     = AjaxGraphQLClient(config.todoURL)
+    val polls    = WebSocketGraphQLClient(config.pollURL)
+  }
+
+  object Views {
+    lazy val target: View[F, Option[Target]]              = rootModel.view(RootModel.target)
+    lazy val persons: View[F, List[TestQuery.AllPersons]] = rootModel.view(RootModel.persons)
+    lazy val todoList: View[F, List[Task]]                = rootModel.view(RootModel.todoList)
+    lazy val polls: View[F, List[Poll]]                   = rootModel.view(RootModel.polls)
+  }
+
+  object Actions {
+    lazy val persons: PersonsActions[F] = new PersonsActionsInterpreter[F](Clients.starWars)
+    lazy val todoList: TodoListActions[F] =
+      new TodoListActionsInterpreter[F](Views.todoList)(Clients.todo)
+    lazy val polls: PollsActions[F] = new PollsActionsInterpreter[F](Views.polls)(Clients.polls)
+  }
+
+  def cleanup(): F[Unit] =
+    // Close the websocket
+    Clients.polls.close()
+}
+
+object AppStateIO {
+  import scala.concurrent.ExecutionContext.global
+
+  implicit lazy val timerIO: Timer[IO]     = cats.effect.IO.timer(global)
+  implicit lazy val csIO: ContextShift[IO] = IO.contextShift(global)
+
+  private val value: Ref[SyncIO, Option[AppState[IO]]] =
+    Ref.unsafe[SyncIO, Option[AppState[IO]]](None)
```

Comment:
Yes, that would be ideal. Let me try to do that. We just have to make sure they remain accesible to components, but they could be moved to the `AppState` class.

Reply:
I took another look at this and the thing is, `Timer` and `ContextShift` are needed to create the `Ref`.

This takes us all the way back to deciding where to keep global application state and action interpreters.

Currently, state and action interpreters are bundled in a single class (`AppState`), but this may be the wrong approach. State could be held by a root component and passed down the component hierarchy as props, but doing this with interpreters seems impractical. We could have the interpreters in a React context, but I'm afraid this may pollute the components with boilerplate.

Also, this `AppState` is defined once in `common` and resides in an object accesible from all the sub-applications. We may want each sub-application to define its part of the state or interpreters that is not shared, in which case we should come up with a way to compose these and make them easily accessible to components.

In any case, I'm looking forward to discussing the issue since I'm not really satisfied with the current implementation.